File: programmers/PART01/메뉴리뉴얼.py
from itertools import combinations


def solution(orders, course):
    answer = []
    '''
    각 주문의 모든 조합을 찾는다. 이 때 메뉴는 2 개 이상이다.
    가장 많은 조합을 순서대로 나열하고 길이가 같은 것을 추가하여 반환한다.
    '''

    count = {}
    for c in course:
        for order in orders:
            order = sorted(order)
            for cases in combinations(order, c):
                cases = ''.join(cases)
                count[cases] = count.get(cases, 0) + 1
    print(count)
    last_len = None
    largest = None
    for menu, m in sorted(count.items(), key=lambda x: x[1], reverse=True):
        '''
        1. 현재 저장된 값이 없으면 가장 최신을 저장하고 마지막 메뉴의 길이와 메뉴 수를 저장
        2. 현재 메뉴의 길이가 마지막 메뉴의 길이와 동일한데 추천 수가 같으면 값 추가
        2. 현재 메뉴의 길이가 마지막 메뉴의 길이가 동일한데 추천 수가 적으면 continue
        3. 현재 메뉴의 길이가 마지막 메뉴의 길이보다 짧아지면 continue
        4. 현재 메뉴의 길이가 마지막 메뉴의 길이보다 길어지면 1번과 동일
        '''
        if m < 2:
            break
        if last_len is None or len(menu) > last_len:
            answer.append(menu)
            last_len = len(menu)
            largest = m
        elif len(menu) == last_len:
            if largest <= m:
                answer.append(menu)
            elif largest > m:
                continue
        # print(menu, m, answer, last_len, largest)

    return answer


# print(solution(["ABCFG", "AC", "CDE", "ACDE", "BCFG", "ACDEH"], [2, 3, 4]))
# print(solution(["ABCDE", "AB", "CD", "ADE", "XYZ", "XYZ", "ACD"], [2, 3, 5]))
print(solution(["XYZ", "XWY", "WXA"], [2, 3, 4]))

File: programmers/PART01/메뉴리뉴얼.js
// https://programmers.co.kr/learn/courses/30/lessons/72411?language=javascript

Object.filter = (obj, predicate) => Object.fromEntries(Object.entries(obj).filter(predicate));
Object.reduce = (obj, callbackFn, init) => Object.entries(obj).reduce(callbackFn, init);

/**
 * course와 함께 orders를 순회하면서 단품 메뉴로 가능한 구성의 조합을 모두 구한다.
 * 구한 조합을 카운팅하여 1 이하의 조합을 필터링
 *
 * 메뉴 조합 중 각 개수 별로 가장 많이 요청된 조합만 남겨두고 전부 없앤다.
 *
 * @param {string[]} orders 손님들이 주문한 단품메뉴
 * @param {number[]} course 단품메뉴의 개수
 * @return {string[]} 추가하게 될 코스요리의 메뉴 구성
 */
function solution(orders, course) {
  const combs = {};
  const maxCombs = course.reduce((result, key) => ({ ...result, [key]: 0 }), {});
  for (const c of course) {
    for (const order of orders) {
      combinations([...order].sort(), c).forEach((comb) => {
        comb = comb.join("");
        combs[comb] = combs[comb] + 1 || 1;
      });
    }
  }
  const filteredCombs = Object.filter(combs, ([key, value]) => {
    maxCombs[key.length] = Math.max(maxCombs[key.length], value);
    return value > 1;
  });
  return Object.reduce(
    filteredCombs,
    (obj, [k, v]) => obj.concat(maxCombs[k.length] === v ? k : []),
    []
  ).sort();
}

function combinations(iterable, r = iterable.length) {
  if (r === 1) return iterable.map((value) => [value]);
  return iterable.reduce((result, value, index, origin) => {
    const rest = origin.slice(index + 1);
    return result.concat(combinations(rest, r - 1).map((comb) => [value, ...comb]));
  }, []);
}

let orders, course;

[orders, course] = [
  ["ABCFG", "AC", "CDE", "ACDE", "BCFG", "ACDEH"],
  [2, 3, 4],
];
console.log(solution(orders, course)); // ["AC", "ACDE", "BCFG", "CDE"]
[orders, course] = [
  ["XYZ", "XWY", "WXA"],
  [2, 3, 4],
];
console.log(solution(orders, course)); // ["AC", "ACDE", "BCFG", "CDE"]
